core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -372,7 +372,7 @@ final class ShuffleBlockFetcherIterator(
     logDebug("Got local blocks in " + Utils.getUsedTimeMs(startTime))
   }
 
-  override def hasNext: Boolean = numBlocksProcessed < numBlocksToFetch
+  override def hasNext: Boolean = !isZombie && (numBlocksProcessed < numBlocksToFetch)
```

Comment:
Is ShuffleBlockFetcherIterator's change related to PipedRDD?

Reply:
Kind of. The root cause of OOM I described in [Spark-26713](https://issues.apache.org/jira/browse/SPARK-26713) is ShuffledRDD + PipedRDD.

I found PipedRDD's stderr writer hangs at `ShuffleBlockFetcherIterator.next()` and leaks memory. I believe this change of `ShuffleBlockFetcherIterator's` could reduce the possibility of race condition and It seems right to mark iterator as fully consumed if is already cleaned up.

Reply:
After the change at PipedRDD by this PR, won't the stdin writer thread (you wrote stderr writer but I think it is typo) be interrupted? If so, it stops consuming `ShuffleBlockFetcherIterator` anymore. Isn't it enough to solve that?

Reply:
> (you wrote stderr writer but I think it is typo) 

Sorry for the typo.
> If so, it stops consuming ShuffleBlockFetcherIterator anymore. Isn't it enough to solve that?

Yes, for the ShuffledRDD + PipedRRDD case , the cleanup logic in PipedRDD is enough to solve the potential leak.
However I am thinking that ShuffledRDD could be transformed with any operations, there may be other cases that `ShuffledBlockFetcherIterator` is cleaned up but still being consumed. So, making the `ShuffledBlockFetcherIterator` defensive.

Reply:
Ok, it sounds good for me. You can make the related comment general and move it to `ShuffledBlockFetcherIterator`.